daemon/emer-circular-file.c

```diff
@@ -0,0 +1,825 @@
+      gsize bytes_start = num_bytes - bytes_head;
+      gboolean read_succeeded =
+        g_input_stream_read_all (input_stream, copy_to, bytes_start,
+                                 NULL /* bytes read */, NULL /* GCancellable */,
+                                 error);
+      if (!read_succeeded)
+        goto handle_failed_read;
+    }
+
+  GSeekable *seekable = G_SEEKABLE (file_input_stream);
+  gboolean seek_succeeded =
+    g_seekable_seek (seekable, priv->head, G_SEEK_SET, NULL /* GCancellable */,
+                     error);
+  if (!seek_succeeded)
+    goto handle_failed_read;
+
+  gboolean read_succeeded =
+    g_input_stream_read_all (input_stream, buffer, bytes_head,
+                             NULL /* bytes read */, NULL /* GCancellable */,
+                             error);
+  g_object_unref (file_input_stream);
+  return read_succeeded;
+
+handle_failed_read:
+  g_object_unref (file_input_stream);
+  return FALSE;
+}
+
+/* Returns the size of a length-encoded buffer excluding any truncated trailing
+ * element. Assumes each element in the buffer is preceded by its length in
+ * bytes encoded as a little-endian guint64. Also assumes that each length,
+ * element pair is concatenated to the next.
+ */
+static guint64
+get_trimmed_size (const guint8 *buffer,
+                  gsize         num_bytes)
+{
+  guint64 curr_pos = 0;
+  while ((curr_pos + sizeof (curr_pos)) < num_bytes)
+    {
+      const guint64 *little_endian_elem_size =
+        (const guint64 *) (buffer + curr_pos);
+      guint64 elem_size =
+        swap_bytes_64_if_big_endian (*little_endian_elem_size);
+      guint64 next_pos = curr_pos + sizeof (curr_pos) + elem_size;
+      if (next_pos > num_bytes)
+        break;
+      curr_pos = next_pos;
+    }
+
+  return curr_pos;
+}
+
+/* Replace the data file with the contents of the given buffer, and update the
+ * metadata file accordingly. In the event of an error, the circular file may
+ * instead be left logically empty, but the data and metadata files will never
+ * be left in an inconsistent state.
+ */
+static gboolean
+overwrite (EmerCircularFile *self,
+           gconstpointer     buffer,
+           gsize             num_bytes,
+           guint64           prev_max_size,
+           GError          **error)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  guint64 little_endian_max_size = swap_bytes_64_if_big_endian (priv->max_size);
+  /* The size of the data file may not exceed the maximum size stored in the
+   * metadata file.
+   */
+  if (priv->max_size > prev_max_size)
+    g_key_file_set_uint64 (priv->metadata_key_file, METADATA_GROUP_NAME,
+                           MAX_SIZE_KEY, little_endian_max_size);
+  if (!set_metadata (self, 0, 0, error))
+    return FALSE;
+
+  gchar *data_filepath = g_file_get_path (priv->data_file);
+  gboolean write_succeeded =
+    g_file_set_contents (data_filepath, buffer, num_bytes, error);
```

Comment:
You could consider using `g_file_replace_contents()` with the `GFile` directly instead.